libvast/vast/table_slice_visit.hpp

```diff
@@ -0,0 +1,61 @@
+/******************************************************************************
+ *                    _   _____   __________                                  *
+ *                   | | / / _ | / __/_  __/     Visibility                   *
+ *                   | |/ / __ |_\ \  / /          Across                     *
+ *                   |___/_/ |_/___/ /_/       Space and Time                 *
+ *                                                                            *
+ * This file is part of VAST. It is subject to the license terms in the       *
+ * LICENSE file found in the top-level directory of this distribution and at  *
+ * http://vast.io/license. No part of VAST, including this file, may be       *
+ * copied, modified, propagated, or distributed except according to the terms *
+ * contained in the LICENSE file.                                             *
+ ******************************************************************************/
+
+#pragma once
+
+#include "vast/die.hpp"
+#include "vast/fbs/table_slice.hpp"
+#include "vast/fwd.hpp"
+
+#include <functional>
+#include <type_traits>
+
+namespace vast {
+
+/// Visits a FlatBuffers table slice to dispatch to its specific encoding.
+/// @param visitor A callable object to dispatch to.
+/// @param x The FlatBuffers root type for table slices.
+/// @note The handler for invalid table slices takes no arguments. If none is
+/// specified, visig aborts when the table slice encoding is invalid.
+template <class Visitor>
+auto visit(Visitor&& visitor, const fbs::TableSlice* x) noexcept(
+  std::conjunction_v<
+    // Check whether the handler for invalid encodings is noexcept-specified,
+    // if and only if it exists.
+    std::disjunction<std::is_nothrow_invocable<Visitor>,
+                     std::negation<std::is_invocable<Visitor>>>,
```

Comment:
This is kinda hard to understand. We have 3 cases here:
|             | throws | nothrow | absent |
| --------- | --------- | ---------- | --------- |
| nothrow_invocable | false | true | false |
| not invokable | false | false | true |
|    or  | false | true | true |

So it is correct. We could think about switching the order, I assume the default will usually be absent.

Reply:
It should be about 50-50. Currently there's lots of missing defaults, but that is only because we necessarily use it outside of `table_slice.cpp`. I still agree with changing the order, as that makes it easier to read in my opinion.

Ideally, we can move the visit function inside `table_slice.cpp` in the future and stop exposing the underlying FlatBuffers table at all.

Reply:
I have reversed the order as suggested and rebased to get rid of all fixup! commits.